cpp/src/arrow/extension_type.cc

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/extension_type.h"
+
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/array.h"
+#include "arrow/type.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+std::string ExtensionType::ToString() const {
+  std::stringstream ss;
+  ss << "extension<" << this->extension_name() << ">";
+  return ss.str();
+}
+
+std::string ExtensionType::name() const { return "extension"; }
+
+void ExtensionArray::SetData(const std::shared_ptr<ArrayData>& data) {
+  this->Array::SetData(data);
+
+  auto storage_data = data->Copy();
+  storage_data->type = (static_cast<const ExtensionType&>(*data->type).storage_type());
+  storage_ = MakeArray(storage_data);
+}
+
+std::unordered_map<std::string, std::shared_ptr<ExtensionType>> g_extension_registry;
```

Comment:
I think this can make extension types inconvenient.

Suppose I want to define a parametric extension type for fixed-size bigints. The type could be `BigIntType(nbytes)` where `nbytes` is arbitrary large (e.g. one may want to represent RSA key parameters). Do I, or the user of my extension type, have to individually register all possible types with different names each time?

I would rather expect to register an "extension type traits" or "extension type adapter" object, and then be able to use multiple extension type instances referring to the same traits / adapter object, if they only differ by their parameters.

IOW the previous design seemed better to me :-)

Reply:
This design doesn't preclude parametric extension types. You only have to register _some_ instance of the type.